control-plane/connect-inject/controllers/endpointsv2/endpoints_controller_test.go

```diff
@@ -0,0 +1,404 @@
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Add the default namespace.
+			ns := corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default"}}
+			// Create fake k8s client
+			k8sObjects := append(tc.k8sObjects(), &ns)
+
+			fakeClient := fake.NewClientBuilder().WithRuntimeObjects(k8sObjects...).Build()
+
+			// Create test consulServer server.
+			testClient := test.TestServerWithMockConnMgrWatcher(t, func(c *testutil.TestServerConfig) {
+				c.Experiments = []string{"resource-apis"}
+			})
+
+			// Create the endpoints controller.
+			ep := &Controller{
+				Client:                fakeClient,
+				Log:                   logrtest.New(t),
+				ConsulServerConnMgr:   testClient.Watcher,
+				AllowK8sNamespacesSet: mapset.NewSetWith("*"),
+				DenyK8sNamespacesSet:  mapset.NewSetWith(),
+			}
+			resourceClient, err := consul.NewResourceServiceClient(ep.ConsulServerConnMgr)
+			require.NoError(t, err)
+
+			namespacedName := types.NamespacedName{
+				Name:      tc.svcName,
+				Namespace: "default",
+			}
+
+			resp, err := ep.Reconcile(context.Background(), ctrl.Request{
+				NamespacedName: namespacedName,
+			})
+			if tc.expErr != "" {
+				require.EqualError(t, err, tc.expErr)
+			} else {
+				require.NoError(t, err)
+			}
+			require.False(t, resp.Requeue)
+
+			// Default if not specified in test
+			if tc.targetConsulNs == "" {
+				tc.targetConsulNs = constants.DefaultConsulNS
+			}
+			if tc.targetConsulPartition == "" {
+				tc.targetConsulPartition = constants.DefaultConsulPartition
+			}
+
+			expectedServiceMatches(t, resourceClient, tc.svcName, tc.targetConsulNs, tc.targetConsulPartition, tc.expectedResource)
+		})
+	}
+}
+
+// TODO(zalimeni)
+//   - "Allow/deny namespaces"
+//   - "With additional meta updated"
+func TestReconcile_UpdateService(t *testing.T) {
+
+}
+
+// TODO(zalimeni)
+//   - "Allow/deny namespaces"
+//   - "Endpoints not found"
+func TestReconcile_DeleteService(t *testing.T) {
+
+}
+
+func expectedServiceMatches(t *testing.T, client pbresource.ResourceServiceClient, name, namespace, partition string, expectedResource *pbresource.Resource) {
+	req := &pbresource.ReadRequest{Id: getServiceID(name, namespace, partition)}
+
+	res, err := client.Read(context.Background(), req)
+
+	if expectedResource == nil {
+		require.Error(t, err)
+		s, ok := status.FromError(err)
+		require.True(t, ok)
+		require.Equal(t, codes.NotFound, s.Code())
+		return
+	}
+
+	require.NoError(t, err)
+	require.NotNil(t, res)
+	require.NotNil(t, res.GetResource().GetData())
+
+	expectedService := &pbcatalog.Service{}
+	err = anypb.UnmarshalTo(expectedResource.Data, expectedService, proto.UnmarshalOptions{})
+	require.NoError(t, err)
+
+	actualService := &pbcatalog.Service{}
+	err = res.GetResource().GetData().UnmarshalTo(actualService)
+	require.NoError(t, err)
+
+	if diff := cmp.Diff(actualService, expectedService, protocmp.Transform()); diff != "" {
+		t.Errorf("unexpected difference:\n%v", diff)
+	}
```

Comment:
@DanStough you might be interested in adding this to your comp fn - it's what we use in goldens and is really nice for a visual diff